diff --git a/dlls/kernelbase/file.c b/dlls/kernelbase/file.c
index f4f9d14518f..69005607224 100644
--- a/dlls/kernelbase/file.c
+++ b/dlls/kernelbase/file.c
@@ -2918,6 +2918,21 @@ BOOL WINAPI DECLSPEC_HOTPATCH SetCurrentDirectoryW( LPCWSTR dir )
 {
     UNICODE_STRING dirW;
 
+    wchar_t str[MAX_PATH];
+    DWORD size = GetEnvironmentVariableW(L"SteamGameId", str, MAX_PATH);
+
+    if (size > 0) {
+        if (wcscmp(str, L"218210") == 0) {
+            SIZE_T len = wcslen(dir);
+
+            if (len > 0 && dir[len - 1] == '.') {
+                WCHAR *p = (WCHAR *)dir + len - 1;
+                *p = '\0';
+                FIXME("%s . fixed\n", debugstr_w(dir));
+            }
+        }
+    }
+
     RtlInitUnicodeString( &dirW, dir );
     return set_ntstatus( RtlSetCurrentDirectory_U( &dirW ));
 }
