From 15deada00f09791f95e47178a92b665b2f5cdedc Mon Sep 17 00:00:00 2001
From: Alexandros Frantzis <alexandros.frantzis@collabora.com>
Date: Mon, 17 Mar 2025 13:51:39 +0200
Subject: [PATCH 024/486] winewayland: Warn about missing clipboard
 functionality.

---
 dlls/winewayland.drv/wayland.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/winewayland.drv/wayland.c b/dlls/winewayland.drv/wayland.c
index 2a51222fd5f..7caf33c872e 100644
--- a/dlls/winewayland.drv/wayland.c
+++ b/dlls/winewayland.drv/wayland.c
@@ -311,6 +311,14 @@ BOOL wayland_process_init(void)
     if (!process_wayland.zwp_text_input_manager_v3)
         ERR("Wayland compositor doesn't support optional zwp_text_input_manager_v3 (host input methods won't work)\n");
 
+    if (!process_wayland.zwlr_data_control_manager_v1)
+    {
+        if (!process_wayland.wl_data_device_manager)
+            ERR("Wayland compositor doesn't support optional wl_data_device_manager (clipboard won't work)\n");
+        else
+            ERR("Wayland compositor doesn't support optional zwlr_data_control_manager_v1 (clipboard functionality will be limited)\n");
+    }
+
     process_wayland.initialized = TRUE;
 
     return TRUE;
-- 
2.53.0

