From fa78747547a94f8cc551993074f7f806775abdb9 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty <etaash.mathamsetty@gmail.com>
Date: Mon, 20 Jan 2025 15:40:46 -0500
Subject: [PATCH 035/486] wine.inf: Don't clobber UBR key.

---
 loader/wine.inf.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 1b822ad1eab..3a0dd59d492 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -2691,7 +2691,7 @@ HKLM,%CurrentVersionNT%,"CurrentMinorVersionNumber",0x10003,0
 HKLM,%CurrentVersionNT%,"CurrentBuild",2,"19045"
 HKLM,%CurrentVersionNT%,"CurrentBuildNumber",2,"19045"
 HKLM,%CurrentVersionNT%,"CurrentType",2,"Multiprocessor Free"
-HKLM,%CurrentVersionNT%,"UBR",0x10001,5796
+HKLM,%CurrentVersionNT%,"UBR",0x10003,5796
 HKLM,%CurrentVersionNT%,"DigitalProductId",2,00,00,00,00,00,00,00,00,00,00,00,\
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,\
 00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,\
-- 
2.53.0

