From 419d82ac381a162b49f56e66a60957da81817107 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty <etaash.mathamsetty@gmail.com>
Date: Wed, 29 Jan 2025 19:05:17 -0500
Subject: [PATCH 036/486] twinapi.appcore/tests: Fix broken registry query.

---
 dlls/twinapi.appcore/tests/twinapi.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/twinapi.appcore/tests/twinapi.c b/dlls/twinapi.appcore/tests/twinapi.c
index 5ee5c45fc91..e86281b895e 100644
--- a/dlls/twinapi.appcore/tests/twinapi.c
+++ b/dlls/twinapi.appcore/tests/twinapi.c
@@ -164,6 +164,7 @@ static void test_AnalyticsVersionInfo(void)
     WindowsDeleteString( str );
     WindowsDeleteString( expect_str );
 
+    size = sizeof(revision);
     if (RegGetValueW( HKEY_LOCAL_MACHINE, L"Software\\Microsoft\\Windows NT\\CurrentVersion", L"UBR",
                       RRF_RT_REG_DWORD, NULL, &revision, &size ))
         revision = 0;
-- 
2.53.0

