From f29eb5d4054f58671847711645315d2b28765cfa Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty <etaash.mathamsetty@gmail.com>
Date: Mon, 14 Jul 2025 22:31:22 -0400
Subject: [PATCH 153/486] ntoskrnl: Fixup control register emulation on amd64.

---
 dlls/ntoskrnl.exe/instr.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/ntoskrnl.exe/instr.c b/dlls/ntoskrnl.exe/instr.c
index 069e348127f..694acadaaaa 100644
--- a/dlls/ntoskrnl.exe/instr.c
+++ b/dlls/ntoskrnl.exe/instr.c
@@ -743,10 +743,10 @@ static DWORD emulate_instruction( EXCEPTION_RECORD *rec, CONTEXT *context )
             TRACE( "mov cr%u,%s at %Ix\n", reg, reg_names[rm], context->Rip );
             switch (reg)
             {
-            case 0: *data = 0x10; break; /* FIXME: set more bits ? */
+            case 0: *data = CR0_PE|CR0_ET|CR0_NE|CR0_WP|CR0_AM|CR0_PG; break;
             case 2: *data = 0; break;
             case 3: *data = 0; break;
-            case 4: *data = 0; break;
+            case 4: *data = 0x20; break;
             case 8: *data = 0; break;
             default: return ExceptionContinueSearch;
             }
-- 
2.53.0

