From 28931d29b28470e76407c6b8643e95675caef4db Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty <etaash.mathamsetty@gmail.com>
Date: Sun, 27 Jul 2025 14:21:10 -0400
Subject: [PATCH 196/486] HACK: kernelbase: --in-process-gpu warframe for
 winewayland.

---
 dlls/kernelbase/process.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/kernelbase/process.c b/dlls/kernelbase/process.c
index 3978d4225c7..19fce170d69 100644
--- a/dlls/kernelbase/process.c
+++ b/dlls/kernelbase/process.c
@@ -626,7 +626,8 @@ static const WCHAR *hack_append_command_line( const WCHAR *cmd )
         {L"launcher_epic.exe", L" --in-process-gpu"}, /* ZZZ EGS */
         {L"Battle.net.exe", L" --in-process-gpu"},
         {L"RSI Launcher.exe", L" --in-process-gpu"},
-        {L"EADesktop.exe", L" --in-process-gpu"}
+        {L"EADesktop.exe", L" --in-process-gpu"},
+        {L"Launcher.exe", L" --in-process-gpu", "230410"} /* Warframe steam */
     };
 
     unsigned int i;
-- 
2.53.0

