From d507a91d49e309358ff9174f824dbf65a6c1d5c8 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty <etaash.mathamsetty@gmail.com>
Date: Wed, 30 Jul 2025 17:12:16 -0400
Subject: [PATCH 198/486] kernelbase: Add TRACE to HeapSummary.

---
 dlls/kernelbase/memory.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/kernelbase/memory.c b/dlls/kernelbase/memory.c
index 02de01bea0d..58ca3b47141 100644
--- a/dlls/kernelbase/memory.c
+++ b/dlls/kernelbase/memory.c
@@ -807,6 +807,8 @@ BOOL WINAPI HeapSummary( HANDLE heap, DWORD flags, LPHEAP_SUMMARY heap_summary )
     SIZE_T uncommitted = 0;
     PROCESS_HEAP_ENTRY entry;
 
+    TRACE("%p %lu %p\n", heap, flags, heap_summary);
+
     if (heap_summary->cb != sizeof(*heap_summary))
     {
         /* needs to be set to the exact size by the caller */
-- 
2.53.0

