From 2bab29cb813e1c2c0315528d822ad05b7ccab855 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty <etaash.mathamsetty@gmail.com>
Date: Wed, 30 Jul 2025 18:17:41 -0400
Subject: [PATCH 199/486] ntdll: Add FIXME to SystemModuleInformation.

---
 dlls/ntdll/unix/system.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/ntdll/unix/system.c b/dlls/ntdll/unix/system.c
index b071ea6be5f..6e186d102f5 100644
--- a/dlls/ntdll/unix/system.c
+++ b/dlls/ntdll/unix/system.c
@@ -3433,6 +3433,8 @@ NTSTATUS WINAPI NtQuerySystemInformation( SYSTEM_INFORMATION_CLASS class,
         ULONG i;
         RTL_PROCESS_MODULES *smi = info;
 
+        FIXME("semi-stub!\n");
+
         len = offsetof( RTL_PROCESS_MODULES, Modules[ARRAY_SIZE(fake_modules)] );
         if (len <= size)
         {
@@ -3768,6 +3770,8 @@ NTSTATUS WINAPI NtQuerySystemInformation( SYSTEM_INFORMATION_CLASS class,
         ULONG i;
         RTL_PROCESS_MODULE_INFORMATION_EX *module_info = info;
 
+        FIXME("semi-stub!\n");
+
         len = sizeof(*module_info) * ARRAY_SIZE(fake_modules) + sizeof(module_info->NextOffset);
         if (len <= size)
         {
-- 
2.53.0

