From 32c716c8b0d7ba0c2885d2525e1509d5a5286581 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty <etaash.mathamsetty@gmail.com>
Date: Fri, 1 Aug 2025 12:53:09 -0400
Subject: [PATCH 200/486] mountmgr.sys: Add stub for StorageDeviceTrimProperty.

---
 dlls/mountmgr.sys/device.c | 27 +++++++++++++++++++++++++++
 include/ntddstor.h         |  6 ++++++
 2 files changed, 33 insertions(+)

diff --git a/dlls/mountmgr.sys/device.c b/dlls/mountmgr.sys/device.c
index fc9102f10ea..3cac68a31d4 100644
--- a/dlls/mountmgr.sys/device.c
+++ b/dlls/mountmgr.sys/device.c
@@ -1621,7 +1621,34 @@ static NTSTATUS query_property( struct disk_device *device, IRP *irp )
         }
         break;
     }
+    case StorageDeviceTrimProperty:
+    {
+        DEVICE_TRIM_DESCRIPTOR *d = irp->AssociatedIrp.SystemBuffer;
+
+        if (irpsp->Parameters.DeviceIoControl.OutputBufferLength < sizeof(STORAGE_DESCRIPTOR_HEADER))
+        {
+            status = STATUS_INVALID_PARAMETER;
+        }
+        else
+        {
+            if (irpsp->Parameters.DeviceIoControl.OutputBufferLength < sizeof(*d))
+            {
+                d->Version = d->Size = sizeof(*d);
+                irp->IoStatus.Information = sizeof(DEVICE_TRIM_DESCRIPTOR);
+            }
+            else
+            {
+                FIXME( "Faking StorageDeviceTrimProperty data.\n" );
 
+                memset( d, 0, sizeof(*d) );
+                d->TrimEnabled = TRUE;
+                d->Version = d->Size = sizeof(*d);
+                irp->IoStatus.Information = sizeof(*d);
+            }
+            status = STATUS_SUCCESS;
+        }
+        break;
+    }
     default:
         FIXME( "Unsupported property %#x\n", query->PropertyId );
         status = STATUS_NOT_SUPPORTED;
diff --git a/include/ntddstor.h b/include/ntddstor.h
index 0cb6d01f9b9..653fbaea42d 100644
--- a/include/ntddstor.h
+++ b/include/ntddstor.h
@@ -266,6 +266,12 @@ typedef struct _DEVICE_SEEK_PENALTY_DESCRIPTOR {
     BOOLEAN                     IncursSeekPenalty;
 } DEVICE_SEEK_PENALTY_DESCRIPTOR, *PDEVICE_SEEK_PENALTY_DESCRIPTOR;
 
+typedef struct _DEVICE_TRIM_DESCRIPTOR {
+    ULONG                       Version;
+    ULONG                       Size;
+    BOOLEAN                     TrimEnabled;
+} DEVICE_TRIM_DESCRIPTOR, *PDEVICE_TRIM_DESCRIPTOR;
+
 typedef struct _STORAGE_DESCRIPTOR_HEADER {
     ULONG                       Version;
     ULONG                       Size;
-- 
2.53.0

