From af78aa655f322b8571745fe8edd3a1e2ebdba99c Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty <etaash.mathamsetty@gmail.com>
Date: Mon, 4 Aug 2025 12:21:10 -0400
Subject: [PATCH 201/486] winewayland: Fix uninitialized variable.

---
 dlls/winewayland.drv/display.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/winewayland.drv/display.c b/dlls/winewayland.drv/display.c
index add844af2f1..2b652b9131c 100644
--- a/dlls/winewayland.drv/display.c
+++ b/dlls/winewayland.drv/display.c
@@ -450,7 +450,7 @@ struct wl_output *wayland_get_best_output_for_rect(const RECT *window_rect)
 {
     struct wayland_output *output;
     struct wl_output *best = NULL;
-    RECT output_rect, temp, intersect = {0};
+    RECT output_rect, temp = {0}, intersect = {0};
     struct wl_array output_info_array;
     struct output_info *output_info;
 
-- 
2.53.0

