From 997ab07432f2cbef6151e85e9dc4a4ac520de595 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Fri, 4 Apr 2025 12:04:02 +0300
Subject: [PATCH 226/486] sane.ds: Fix DC leak.

Signed-off-by: Dmitry Timoshkov <dmitry@baikal.ru>
---
 dlls/sane.ds/ui.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/dlls/sane.ds/ui.c b/dlls/sane.ds/ui.c
index 557cf81e0b1..25a1cf33970 100644
--- a/dlls/sane.ds/ui.c
+++ b/dlls/sane.ds/ui.c
@@ -475,8 +475,6 @@ BOOL DoScannerUI(void)
     LPWSTR szCaption;
     DWORD len;
 
-    hdc = GetDC(0);
-
     memset(psp,0,sizeof(psp));
     rc = sane_option_get_value( 0, &optcount );
     if (rc != TWCC_SUCCESS)
@@ -485,6 +483,8 @@ BOOL DoScannerUI(void)
         return FALSE;
     }
 
+    hdc = CreateCompatibleDC(0);
+
     while (index < optcount)
     {
         struct option_descriptor opt;
@@ -539,7 +539,9 @@ BOOL DoScannerUI(void)
         free((LPBYTE)psp[index].pszTitle);
     }
     free(szCaption);
-    
+
+    DeleteDC(hdc);
+
     if (psrc == IDOK)
         return TRUE;
     else
-- 
2.53.0

