From 652aafb7071ad750a92c1bb9fac094c6afe2ddfc Mon Sep 17 00:00:00 2001
From: Anders Kjersem <andersdev@proton.me>
Date: Wed, 19 Mar 2025 23:51:42 +0000
Subject: [PATCH 259/486] shcore: Implement OS_TABLETPC and OS_MEDIACENTER.

---
 dlls/shcore/main.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/dlls/shcore/main.c b/dlls/shcore/main.c
index 58ddda3fb2a..426a842ed28 100644
--- a/dlls/shcore/main.c
+++ b/dlls/shcore/main.c
@@ -2482,14 +2482,12 @@ BOOL WINAPI IsOS(DWORD feature)
     case OS_SMALLBUSINESSSERVER:
         ISOS_RETURN(platform == VER_PLATFORM_WIN32_NT);
     case OS_TABLETPC:
-        FIXME("(OS_TABLETPC) What should we return here?\n");
-        return FALSE;
+        ISOS_RETURN(GetSystemMetrics(SM_TABLETPC));
     case OS_SERVERADMINUI:
         FIXME("(OS_SERVERADMINUI) What should we return here?\n");
         return FALSE;
     case OS_MEDIACENTER:
-        FIXME("(OS_MEDIACENTER) What should we return here?\n");
-        return FALSE;
+        ISOS_RETURN(GetSystemMetrics(SM_MEDIACENTER));
     case OS_APPLIANCE:
         FIXME("(OS_APPLIANCE) What should we return here?\n");
         return FALSE;
-- 
2.53.0

