From de43c19d5d58e2a1f79e40a77d8b37fcb2bc061f Mon Sep 17 00:00:00 2001
From: Tobias Gruetzmacher <tobias-git@23.gs>
Date: Thu, 3 Apr 2025 18:06:49 +0200
Subject: [PATCH 292/486] urlmon: Handle HTTP status code 308 (Permanent
 Redirect).

---
 dlls/urlmon/http.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/urlmon/http.c b/dlls/urlmon/http.c
index a24006ba08c..e592f697485 100644
--- a/dlls/urlmon/http.c
+++ b/dlls/urlmon/http.c
@@ -495,6 +495,7 @@ static BOOL is_redirect_response(DWORD status_code)
     case HTTP_STATUS_MOVED:
     case HTTP_STATUS_REDIRECT_KEEP_VERB:
     case HTTP_STATUS_REDIRECT_METHOD:
+    case HTTP_STATUS_PERMANENT_REDIRECT:
         return TRUE;
     }
     return FALSE;
-- 
2.53.0

