From dd3d59343758ad2b21938957c66ac72ed0591598 Mon Sep 17 00:00:00 2001
From: Herman Semenov <GermanAizek@yandex.ru>
Date: Sun, 19 Jan 2025 13:13:32 +0300
Subject: [PATCH 335/486] msvcrt: Add missing TRACE_ON check.

---
 dlls/msvcrt/undname.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/msvcrt/undname.c b/dlls/msvcrt/undname.c
index 923ec67d363..054bb6b7426 100644
--- a/dlls/msvcrt/undname.c
+++ b/dlls/msvcrt/undname.c
@@ -198,6 +198,7 @@ static BOOL str_array_push(struct parsed_symbol* sym, const char* ptr, int len,
     memcpy(a->elts[a->num], ptr, len);
     a->elts[a->num][len] = '\0'; 
     if (++a->num >= a->max) a->max = a->num;
+    if (TRACE_ON(msvcrt))
     {
         int i;
         char c;
-- 
2.53.0

