From eeee4cde595761bc46f29824cbd1e17509576339 Mon Sep 17 00:00:00 2001
From: Alexandre Julliard <julliard@winehq.org>
Date: Thu, 27 Feb 2025 11:01:48 +0100
Subject: [PATCH 347/486] makefiles: Add support for optionally installing
 external libs.

---
 tools/makedep.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/tools/makedep.c b/tools/makedep.c
index 4ab60eb2797..16cf7491b72 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -3673,9 +3673,8 @@ static void output_static_lib( struct makefile *make, unsigned int arch )
     if (hybrid_arch) output_filenames_obj_dir( make, make->object_files[hybrid_arch] );
     if (!arch) output_filenames_obj_dir( make, make->unixobj_files );
     output( "\n" );
-    if (!make->extlib)
-        add_install_rule( make, make->staticlib, arch, name,
-                          strmake( "d%s%s", arch_install_dirs[arch], make->staticlib ));
+    add_install_rule( make, make->staticlib, arch, name,
+                      strmake( "d%s%s", arch_install_dirs[arch], make->staticlib ));
 }
 
 
@@ -4505,7 +4504,7 @@ static void load_sources( struct makefile *make )
     {
         /* add default install rules if nothing was specified */
         for (i = 0; i < NB_INSTALL_RULES; i++) if (make->install[i].count) break;
-        if (i == NB_INSTALL_RULES)
+        if (i == NB_INSTALL_RULES && !make->extlib)
         {
             if (make->importlib) strarray_add( &make->install[INSTALL_DEV], make->importlib );
             if (make->staticlib) strarray_add( &make->install[INSTALL_DEV], make->staticlib );
-- 
2.53.0

