From 0bf980bc07dfe29aaa98ef485c096eb785174b80 Mon Sep 17 00:00:00 2001
From: Sven Baars <sbaars@codeweavers.com>
Date: Tue, 25 Mar 2025 20:57:36 +0100
Subject: [PATCH 348/486] makedep: Fix a compilation warning.

---
 tools/makedep.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/tools/makedep.c b/tools/makedep.c
index 16cf7491b72..9c7ac1992e8 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -526,6 +526,7 @@ static const char *get_root_relative_path( struct makefile *make )
         while (*dir && *dir != '/') dir++;
         while (*dir == '/') dir++;
     }
+    assert(dotdots);
     ret = xmalloc( 3 * dotdots );
     for (p = ret; dotdots; dotdots--, p += 3) memcpy( p, "../", 3 );
     p[-1] = 0;  /* remove trailing slash */
-- 
2.53.0

