From ff18e964372d44f7ae19c349d7ff4c8137314149 Mon Sep 17 00:00:00 2001
From: Alexandre Julliard <julliard@winehq.org>
Date: Thu, 27 Feb 2025 11:04:53 +0100
Subject: [PATCH 351/486] makefiles: Link to compiler-rt for builds in MSVC
 mode.

---
 tools/makedep.c         | 4 ++++
 tools/winegcc/winegcc.c | 2 ++
 2 files changed, 6 insertions(+)

diff --git a/tools/makedep.c b/tools/makedep.c
index 9c7ac1992e8..6acd0fb8b00 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -159,6 +159,7 @@ static const char *ln_s;
 static const char *sed_cmd;
 static const char *wayland_scanner;
 static const char *sarif_converter;
+static const char *compiler_rt;
 static int so_dll_supported;
 static int unix_lib_supported;
 /* per-architecture global variables */
@@ -2301,6 +2302,7 @@ static struct strarray get_default_imports( const struct makefile *make, struct
         for (i = 0; i < imports.count; i++)
             if (!strcmp( imports.str[i], "winecrt0" )) return ret;
         strarray_add( &ret, "winecrt0" );
+        if (compiler_rt) strarray_add( &ret, compiler_rt );
         return ret;
     }
 
@@ -2309,6 +2311,7 @@ static struct strarray get_default_imports( const struct makefile *make, struct
             crt_dll = imports.str[i];
 
     strarray_add( &ret, "winecrt0" );
+    if (compiler_rt) strarray_add( &ret, compiler_rt );
     if (crt_dll) strarray_add( &ret, crt_dll );
 
     if (make->is_win16 && (!make->importlib || strcmp( make->importlib, "kernel" )))
@@ -4698,6 +4701,7 @@ int main( int argc, char *argv[] )
     ln_s               = get_expanded_make_variable( top_makefile, "LN_S" );
     wayland_scanner    = get_expanded_make_variable( top_makefile, "WAYLAND_SCANNER" );
     sarif_converter    = get_expanded_make_variable( top_makefile, "SARIF_CONVERTER" );
+    compiler_rt        = get_expanded_make_variable( top_makefile, "COMPILER_RT_PE_LIBS" );
 
     if (root_src_dir && !strcmp( root_src_dir, "." )) root_src_dir = NULL;
     if (tools_dir && !strcmp( tools_dir, "." )) tools_dir = NULL;
diff --git a/tools/winegcc/winegcc.c b/tools/winegcc/winegcc.c
index 2421bbe3fc8..622a769839c 100644
--- a/tools/winegcc/winegcc.c
+++ b/tools/winegcc/winegcc.c
@@ -1380,6 +1380,8 @@ static void build(struct options* opts)
         add_library(opts, lib_dirs, &files, "advapi32");
         add_library(opts, lib_dirs, &files, "user32");
         add_library(opts, lib_dirs, &files, "winecrt0");
+        if (opts->target.platform == PLATFORM_WINDOWS)
+            add_library(opts, lib_dirs, &files, "compiler-rt");
         if (opts->use_msvcrt)
         {
             if (!crt_lib)
-- 
2.53.0

