From 81f5355cc2dd68732077398d08a74054285bcee1 Mon Sep 17 00:00:00 2001
From: Alexandre Julliard <julliard@winehq.org>
Date: Fri, 7 Mar 2025 15:48:16 +0100
Subject: [PATCH 354/486] winegcc: PE targets always use msvcrt mode, remove
 redundant checks.

---
 tools/winegcc/winegcc.c | 56 +++++++++++++----------------------------
 1 file changed, 17 insertions(+), 39 deletions(-)

diff --git a/tools/winegcc/winegcc.c b/tools/winegcc/winegcc.c
index a2144a4af7b..facf734a9f8 100644
--- a/tools/winegcc/winegcc.c
+++ b/tools/winegcc/winegcc.c
@@ -587,8 +587,8 @@ static struct strarray get_link_args( struct options *opts, const char *output_n
         else strarray_add( &flags, opts->gui_app ? "-mwindows" : "-mconsole" );
 
         if (opts->unicode_app) strarray_add( &flags, "-municode" );
-        if (opts->nodefaultlibs || opts->use_msvcrt) strarray_add( &flags, "-nodefaultlibs" );
-        if (opts->nostartfiles || opts->use_msvcrt) strarray_add( &flags, "-nostartfiles" );
+        strarray_add( &flags, "-nodefaultlibs" );
+        strarray_add( &flags, "-nostartfiles" );
         if (opts->subsystem) strarray_add( &flags, strmake("-Wl,--subsystem,%s", opts->subsystem ));
 
         strarray_add( &flags, "-Wl,--exclude-all-symbols" );
@@ -601,10 +601,6 @@ static struct strarray get_link_args( struct options *opts, const char *output_n
         if (opts->large_address_aware && opts->target.cpu == CPU_i386)
             strarray_add( &flags, "-Wl,--large-address-aware" );
 
-        /* make sure we don't need a libgcc_s dll on Windows */
-        if (!opts->nodefaultlibs && !opts->use_msvcrt)
-            strarray_add( &flags, "-static-libgcc" );
-
         if (opts->debug_file && strendswith(opts->debug_file, ".pdb"))
             strarray_add(&link_args, strmake("-Wl,--pdb=%s", opts->debug_file));
 
@@ -633,9 +629,9 @@ static struct strarray get_link_args( struct options *opts, const char *output_n
             strarray_add( &flags, "-Wl,-kill-at" );
         }
         if (opts->unicode_app) strarray_add( &flags, "-municode" );
-        if (opts->nodefaultlibs || opts->use_msvcrt) strarray_add( &flags, "-nodefaultlibs" );
+        strarray_add( &flags, "-nodefaultlibs" );
+        strarray_add( &flags, "-nostdlib" );
         if (opts->nostartfiles) strarray_add( &flags, "-nostartfiles" );
-        if (opts->use_msvcrt) strarray_add( &flags, "-nostdlib" );
         if (opts->image_base) strarray_add( &flags, strmake("-Wl,-base:%s", opts->image_base ));
         if (opts->subsystem)
             strarray_add( &flags, strmake("-Wl,-subsystem:%s", opts->subsystem ));
@@ -1433,18 +1429,15 @@ static void build(struct options* opts)
     /* link everything together now */
     link_args = get_link_args( opts, output_name );
 
-    if (opts->nodefaultlibs || opts->use_msvcrt)
+    switch (opts->target.platform)
     {
-        switch (opts->target.platform)
-        {
-        case PLATFORM_MINGW:
-        case PLATFORM_CYGWIN:
-            libgcc = find_libgcc( opts->prefix, link_args );
-            if (!libgcc) libgcc = "-lgcc";
-            break;
-        default:
-            break;
-        }
+    case PLATFORM_MINGW:
+    case PLATFORM_CYGWIN:
+        libgcc = find_libgcc( opts->prefix, link_args );
+        if (!libgcc) libgcc = "-lgcc";
+        break;
+    default:
+        break;
     }
 
     strarray_add(&link_args, "-o");
@@ -1453,7 +1446,7 @@ static void build(struct options* opts)
     for ( j = 0; j < lib_dirs.count; j++ )
 	strarray_add(&link_args, strmake("-L%s", lib_dirs.str[j]));
 
-    if (is_pe && opts->use_msvcrt && !entry_point && (opts->shared || opts->win16_app))
+    if (is_pe && !entry_point && (opts->shared || opts->win16_app))
         entry_point = opts->target.cpu == CPU_i386 ? "DllMainCRTStartup@12" : "DllMainCRTStartup";
 
     if (is_pe && entry_point)
@@ -1494,27 +1487,12 @@ static void build(struct options* opts)
 	    case 'a':
                 if (!opts->use_msvcrt && !opts->lib_suffix && strchr(name, '/'))
                 {
-                    /* turn the path back into -Ldir -lfoo options
-                     * this makes sure that we use the specified libs even
-                     * when mingw adds its own import libs to the link */
                     const char *p = get_basename( name );
 
-                    if (is_pe)
-                    {
-                        if (!strncmp( p, "lib", 3 ) && strcmp( p, "libmsvcrt.a" ))
-                        {
-                            strarray_add(&link_args, strmake("-L%s", get_dirname(name) ));
-                            strarray_add(&link_args, strmake("-l%s", get_basename_noext( p + 3 )));
-                            break;
-                        }
-                    }
-                    else
-                    {
-                        /* don't link to ntdll or ntoskrnl in non-msvcrt mode
-                         * since they export CRT functions */
-                        if (!strcmp( p, "libntdll.a" )) break;
-                        if (!strcmp( p, "libntoskrnl.a" )) break;
-                    }
+                    /* don't link to ntdll or ntoskrnl in non-msvcrt mode
+                     * since they export CRT functions */
+                    if (!strcmp( p, "libntdll.a" )) break;
+                    if (!strcmp( p, "libntoskrnl.a" )) break;
                 }
 		strarray_add(&link_args, name);
 		break;
-- 
2.53.0

