From 36d4fe419e034b63480a62123385e9007302d216 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Fri, 7 Feb 2025 09:53:48 +0800
Subject: [PATCH 448/486] ntdll: Implement RtlLookupElementGenericTable().

---
 dlls/ntdll/rtl.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/dlls/ntdll/rtl.c b/dlls/ntdll/rtl.c
index 001bf6d77ae..fdeb3b2c9c6 100644
--- a/dlls/ntdll/rtl.c
+++ b/dlls/ntdll/rtl.c
@@ -655,10 +655,18 @@ void * WINAPI RtlGetElementGenericTable(RTL_GENERIC_TABLE *table, ULONG index)
 /******************************************************************************
  *  RtlLookupElementGenericTable           [NTDLL.@]
  */
-void * WINAPI RtlLookupElementGenericTable(RTL_GENERIC_TABLE *table, void *buffer)
+void * WINAPI RtlLookupElementGenericTable(RTL_GENERIC_TABLE *table, void *value)
 {
-    FIXME("(%p, %p) stub!\n", table, buffer);
-    return NULL;
+    RTL_SPLAY_LINKS *child;
+
+    TRACE("(%p, %p)\n", table, value);
+
+    child = rtl_splay_find(table, value);
+    if (!child)
+        return FALSE;
+
+    table->TableRoot = RtlSplay(child);
+    return get_data_from_splay_links(child);
 }
 
 /******************************************************************************
-- 
2.53.0

