From 776ee001e29a0c659ab71314f5662120e4434a4e Mon Sep 17 00:00:00 2001
From: Ziia Shi <mkrsym1@gmail.com>
Date: Tue, 3 Feb 2026 20:54:24 +0100
Subject: [PATCH 484/486] win32u: Call SetIMEEnabled from
 NtUserAssociateInputContext.

---
 dlls/win32u/imm.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/win32u/imm.c b/dlls/win32u/imm.c
index 1c677772e0c..3858a9f092d 100644
--- a/dlls/win32u/imm.c
+++ b/dlls/win32u/imm.c
@@ -223,6 +223,9 @@ UINT WINAPI NtUserAssociateInputContext( HWND hwnd, HIMC ctx, ULONG flags )
     }
 
     release_win_ptr( win );
+
+    user_driver->pSetIMEEnabled( hwnd, ctx != NULL );
+
     return ret;
 }
 
-- 
2.53.0

