From 46428d4b4febb0218c68abf863f251b6d1e866eb Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Thu, 18 Dec 2025 16:30:16 -0600
Subject: [PATCH] fixup! ntdll: Wait for thread suspension in
 NtSuspendThread().

---
 dlls/ntdll/unix/thread.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/ntdll/unix/thread.c b/dlls/ntdll/unix/thread.c
index 54b0420d2f1..10f1fb3df51 100644
--- a/dlls/ntdll/unix/thread.c
+++ b/dlls/ntdll/unix/thread.c
@@ -1690,6 +1690,8 @@ NTSTATUS WINAPI NtSuspendThread( HANDLE handle, ULONG *ret_count )
     if (ret == STATUS_PENDING && wait_handle)
     {
         NtWaitForSingleObject( wait_handle, FALSE, NULL );
+        /* remove the handle from the cache, get_thread_context will close it for us */
+        close_inproc_sync_obj( wait_handle );
 
         SERVER_START_REQ( suspend_thread )
         {
-- 
2.52.0

